fish-rust/src/event.rs

```diff
@@ -0,0 +1,658 @@
+use autocxx::WithinUniquePtr;
+use cxx::SharedPtr;
+use libc::{pid_t, SIGWINCH};
+use std::sync::{
+    atomic::{AtomicBool, AtomicU32, Ordering},
+    Arc, Mutex,
+};
+use widestring_suffix::widestrs;
+
+use crate::builtins::shared::io_streams_t;
+use crate::common::{escape_string, EscapeFlags, EscapeStringStyle, ScopedPush};
+use crate::ffi::{
+    block_t, block_type_t, event_block_list_blocks_type, io_chain_t, parser_t, signal_check_cancel,
+    signal_handle, termsize_container_t, Repin,
+};
+use crate::flog::FLOG;
+use crate::signal::{sig2wcs, signal_get_desc};
+use crate::wchar::L;
+use crate::wchar_ffi::WCharToFFI;
+use crate::wchar_ffi::{wstr, WString};
+use crate::wutil::sprintf;
+
+#[cxx::bridge]
+mod ffi {
+    extern "Rust" {
+        type BlockedEvents;
+        type Event;
+    }
+}
```

Comment:
Nice work! The only thing that was missing was the `cxx::bridge` to make sure C++ can talk to the new module.
I usually try to define that before implementing any function bodies.

Here's a fairly minimal diff that makes it compile.

```diff
diff --git a/fish-rust/build.rs b/fish-rust/build.rs
index c485a3374..0c6f42f9f 100644
--- a/fish-rust/build.rs
+++ b/fish-rust/build.rs
@@ -19,6 +19,7 @@ fn main() -> miette::Result<()> {
     // This allows "Rust to be used from C++"
     // This must come before autocxx so that cxx can emit its cxx.h header.
     let source_files = vec![
+        "src/event.rs",
         "src/fd_readable_set.rs",
         "src/ffi_init.rs",
         "src/ffi_tests.rs",
diff --git a/fish-rust/src/builtins/shared.rs b/fish-rust/src/builtins/shared.rs
index 6fb7ec1be..22dffc746 100644
--- a/fish-rust/src/builtins/shared.rs
+++ b/fish-rust/src/builtins/shared.rs
@@ -60,18 +60,18 @@ pub struct io_streams_t {
 }
 
 impl io_streams_t {
-    fn new(mut streams: Pin<&mut builtins_ffi::io_streams_t>) -> io_streams_t {
+    pub fn new(mut streams: Pin<&mut builtins_ffi::io_streams_t>) -> io_streams_t {
         let out = output_stream_t(streams.as_mut().get_out().unpin());
         let err = output_stream_t(streams.as_mut().get_err().unpin());
         let streams = streams.unpin();
         io_streams_t { streams, out, err }
     }
 
-    fn ffi_pin(&mut self) -> Pin<&mut builtins_ffi::io_streams_t> {
+    pub fn ffi_pin(&mut self) -> Pin<&mut builtins_ffi::io_streams_t> {
         unsafe { Pin::new_unchecked(&mut *self.streams) }
     }
 
-    fn ffi_ref(&self) -> &builtins_ffi::io_streams_t {
+    pub fn ffi_ref(&self) -> &builtins_ffi::io_streams_t {
         unsafe { &*self.streams }
     }
 }
diff --git a/fish-rust/src/event.rs b/fish-rust/src/event.rs
index 85906e5ba..38253787f 100644
--- a/fish-rust/src/event.rs
+++ b/fish-rust/src/event.rs
@@ -1,33 +1,104 @@
+//! Functions for handling event triggers
+//!
+//! Because most of these functions can be called by signal handler, it is important to make it well
+//! defined when these functions produce output or perform memory allocations, since such functions
+//! may not be safely called by signal handlers.
+
+// TODO
+#![allow(unused_imports)]
+#![allow(unused_variables)]
+
 use autocxx::WithinUniquePtr;
-use cxx::SharedPtr;
+use cxx::{CxxVector, CxxWString, SharedPtr, UniquePtr};
 use libc::{pid_t, SIGWINCH};
-use std::sync::{
-    atomic::{AtomicBool, AtomicU32, Ordering},
-    Arc, Mutex,
+use std::{
+    pin::Pin,
+    sync::{
+        atomic::{AtomicBool, AtomicU32, Ordering},
+        Arc, Mutex,
+    },
 };
 use widestring_suffix::widestrs;
 
 use crate::builtins::shared::io_streams_t;
 use crate::common::{escape_string, EscapeFlags, EscapeStringStyle, ScopedPush};
 use crate::ffi::{
-    block_t, block_type_t, event_block_list_blocks_type, io_chain_t, parser_t, signal_check_cancel,
-    signal_handle, termsize_container_t, Repin,
+    self, block_type_t, event_block, event_block_list_blocks_type, io_chain_t, parser_t,
+    signal_check_cancel, signal_handle, termsize_container_t, Repin,
 };
 use crate::flog::FLOG;
 use crate::signal::{sig2wcs, signal_get_desc};
 use crate::wchar::L;
-use crate::wchar_ffi::WCharToFFI;
-use crate::wchar_ffi::{wstr, WString};
+use crate::wchar_ffi::{wcharz_t, wstr, WCharFromFFI, WCharToFFI, WString};
 use crate::wutil::sprintf;
 
 #[cxx::bridge]
-mod ffi {
+mod event_ffi {
+    extern "C++" {
+        include!("wutil.h");
+        include!("parser.h");
+        include!("io.h");
+        type wcharz_t = super::wcharz_t;
+        type parser_t = super::parser_t;
+        type io_streams_t = super::ffi::io_streams_t;
+    }
+    enum event_type_t {
+        any,
+        signal,
+        variable,
+        process_exit,
+        job_exit,
+        caller_exit,
+        generic,
+    }
+
+    struct event_description_t {
+        typ: event_type_t,
+        signal: i32,
+        pid: i32,
+        internal_job_id: u64,
+        caller_id: u64,
+        str_param1: UniquePtr<CxxWString>,
+    }
+
     extern "Rust" {
         type BlockedEvents;
         type Event;
+        type EventHandler;
+
+        fn new_event_generic(desc: wcharz_t) -> Box<Event>;
+        fn new_event_variable_erase(name: &CxxWString) -> Box<Event>;
+        fn new_event_variable_set(name: &CxxWString) -> Box<Event>;
+        fn new_event_process_exit(pid: i32, status: i32) -> Box<Event>;
+        fn new_event_job_exit(pgid: i32, jid: u64) -> Box<Event>;
+        fn new_event_caller_exit(internal_job_id: u64, job_id: i32) -> Box<Event>;
+
+        fn event_add_handler(desc: &event_description_t, name: &CxxWString);
+        fn desc(self: &EventHandler) -> event_description_t;
+        fn function_name(self: &EventHandler) -> UniquePtr<CxxWString>;
+        fn set_removed(self: &mut EventHandler);
+
+        fn event_fire_generic_ffi(
+            parser: Pin<&mut parser_t>,
+            name: &CxxWString,
+            arguments: &CxxVector<wcharz_t>,
+        );
+        #[cxx_name = "event_get_desc"]
+        fn event_get_desc_ffi(parser: &parser_t, evt: &Event) -> UniquePtr<CxxWString>;
+        #[cxx_name = "event_fire_delayed"]
+        fn event_fire_delayed_ffi(parser: Pin<&mut parser_t>);
+        #[cxx_name = "event_fire"]
+        fn event_fire_ffi(parser: Pin<&mut parser_t>, event: &Event);
+        #[cxx_name = "event_print"]
+        fn event_print_ffi(streams: Pin<&mut io_streams_t>, type_filter: &CxxWString);
+
+        fn enqueue_signal(signal: usize);
+        fn is_signal_observed(sig: usize) -> bool;
     }
 }
 
+pub use event_ffi::{event_description_t, event_type_t};
+
 struct BlockedEvents(Vec<Arc<Event>>);
 
 const ANY_PID: pid_t = 0;
@@ -99,13 +170,87 @@ impl EventType {
     }
 }
 
+impl From<&EventType> for event_type_t {
+    fn from(typ: &EventType) -> Self {
+        match typ {
+            EventType::Any => event_type_t::any,
+            EventType::Signal { .. } => event_type_t::signal,
+            EventType::Variable { .. } => event_type_t::variable,
+            EventType::ProcessExit { .. } => event_type_t::process_exit,
+            EventType::JobExit { .. } => event_type_t::job_exit,
+            EventType::CallerExit { .. } => event_type_t::caller_exit,
+            EventType::Generic { .. } => event_type_t::generic,
+        }
+    }
+}
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct EventDescription {
     typ: EventType,
 }
 
+impl From<&event_description_t> for EventDescription {
+    fn from(desc: &event_description_t) -> Self {
+        EventDescription {
+            typ: match desc.typ {
+                event_type_t::any => EventType::Any,
+                event_type_t::signal => EventType::Signal {
+                    signal: desc.signal.try_into().unwrap(),
+                },
+                event_type_t::variable => EventType::Variable {
+                    name: desc.str_param1.from_ffi(),
+                },
+                event_type_t::process_exit => EventType::ProcessExit { pid: desc.pid },
+                event_type_t::job_exit => EventType::JobExit {
+                    pid: desc.pid,
+                    internal_job_id: desc.internal_job_id,
+                },
+                event_type_t::caller_exit => EventType::CallerExit {
+                    caller_id: desc.caller_id,
+                },
+                event_type_t::generic => EventType::Generic {
+                    param: desc.str_param1.from_ffi(),
+                },
+                _ => panic!("invalid event description"),
+            },
+        }
+    }
+}
+
+impl From<&EventDescription> for event_description_t {
+    fn from(desc: &EventDescription) -> Self {
+        let mut result = event_description_t {
+            typ: (&desc.typ).into(),
+            signal: Default::default(),
+            pid: Default::default(),
+            internal_job_id: Default::default(),
+            caller_id: Default::default(),
+            str_param1: match desc.typ.str_param1() {
+                Some(param) => param.to_ffi(),
+                None => UniquePtr::null(),
+            },
+        };
+        match desc.typ {
+            EventType::Any => (),
+            EventType::Signal { signal } => result.signal = signal.try_into().unwrap(),
+            EventType::Variable { .. } => (),
+            EventType::ProcessExit { pid } => result.pid = pid,
+            EventType::JobExit {
+                pid,
+                internal_job_id,
+            } => {
+                result.pid = pid;
+                result.internal_job_id = internal_job_id;
+            }
+            EventType::CallerExit { caller_id } => result.caller_id = caller_id,
+            EventType::Generic { .. } => (),
+        }
+        result
+    }
+}
+
 #[derive(Debug)]
-struct EventHandler {
+pub struct EventHandler {
     desc: EventDescription,
     function_name: WString,
     removed: AtomicBool,
@@ -181,6 +326,18 @@ impl EventHandler {
 }
 type EventHandlerList = Vec<Arc<EventHandler>>;
 
+impl EventHandler {
+    fn desc(&self) -> event_description_t {
+        (&self.desc).into()
+    }
+    fn function_name(self: &EventHandler) -> UniquePtr<CxxWString> {
+        self.function_name.to_ffi()
+    }
+    fn set_removed(self: &mut EventHandler) {
+        self.removed.store(true, Ordering::SeqCst);
+    }
+}
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct Event {
     desc: EventDescription,
@@ -188,6 +345,15 @@ pub struct Event {
 }
 
 impl Event {
+    pub fn generic(desc: WString) -> Self {
+        Self {
+            desc: EventDescription {
+                typ: EventType::Generic { param: desc },
+            },
+            arguments: vec![],
+        }
+    }
+
     pub fn variable_erase(name: WString) -> Self {
         Self {
             desc: EventDescription {
@@ -257,15 +423,46 @@ impl Event {
                 break;
             };
 
-            if event_block_list_blocks_type(block.event_blocks()) {
+            if event_block_list_blocks_type(unsafe { &*block.event_blocks() }) {
                 return true;
             }
         }
 
-        event_block_list_blocks_type(&parser.global_event_blocks())
+        event_block_list_blocks_type(unsafe { &*parser.global_event_blocks() })
     }
 }
 
+fn new_event_generic(desc: wcharz_t) -> Box<Event> {
+    Box::new(Event::generic(desc.into()))
+}
+
+fn new_event_variable_erase(name: &CxxWString) -> Box<Event> {
+    Box::new(Event::variable_erase(name.from_ffi()))
+}
+
+fn new_event_variable_set(name: &CxxWString) -> Box<Event> {
+    Box::new(Event::variable_erase(name.from_ffi()))
+}
+
+fn new_event_process_exit(pid: i32, status: i32) -> Box<Event> {
+    Box::new(Event::process_exit(pid, status))
+}
+
+fn new_event_job_exit(pgid: i32, jid: u64) -> Box<Event> {
+    Box::new(Event::job_exit(pgid, jid))
+}
+
+fn new_event_caller_exit(internal_job_id: u64, job_id: i32) -> Box<Event> {
+    Box::new(Event::caller_exit(internal_job_id, job_id))
+}
+
+fn event_add_handler(desc: &event_description_t, name: &CxxWString) {
+    add_handler(Arc::new(EventHandler::new(
+        desc.into(),
+        Some(name.from_ffi()),
+    )));
+}
+
 const SIGNAL_COUNT: usize = 65; // FIXME: NSIG
 
 struct PendingSignals {
@@ -350,6 +547,8 @@ fn dec_signal_observed(sig: usize) {
     }
 }
 
+/// Returns whether an event listener is registered for the given signal. This is safe to call from
+/// a signal handler.
 pub fn is_signal_observed(sig: usize) -> bool {
     // We are in a signal handler!
     if let Some(sig) = s_observed_signals.get(sig) {
@@ -359,7 +558,7 @@ pub fn is_signal_observed(sig: usize) -> bool {
     }
 }
 
-pub fn event_get_desc(parser: &mut parser_t, evt: &Event) -> WString {
+pub fn event_get_desc(parser: &parser_t, evt: &Event) -> WString {
     let s = match &evt.desc.typ {
         EventType::Signal { signal } => format!(
             "signal handler for {} ({})",
@@ -389,6 +588,11 @@ pub fn event_get_desc(parser: &mut parser_t, evt: &Event) -> WString {
     WString::from_str(&s)
 }
 
+fn event_get_desc_ffi(parser: &parser_t, evt: &Event) -> UniquePtr<CxxWString> {
+    event_get_desc(parser, evt).to_ffi()
+}
+
+/// Add an event handler.
 pub fn add_handler(eh: Arc<EventHandler>) {
     if let EventType::Signal { signal } = eh.desc.typ {
         signal_handle(
@@ -485,7 +689,10 @@ fn fire_internal(parser: &mut parser_t, event: &Event) {
         // Event handlers are not part of the main flow of code, so they are marked as
         // non-interactive.
         let _interactive = ScopedPush::new(&mut ld.is_interactive, false);
-        let prev_statuses = parser.get_last_statuses().within_unique_ptr();
+        if true {
+            todo!();
+        }
+        // let prev_statuses = parser.get_last_statuses().within_unique_ptr();
 
         FLOG!(
             event,
@@ -496,16 +703,18 @@ fn fire_internal(parser: &mut parser_t, event: &Event) {
             "'"
         );
 
-        let parser = parser.pin();
-        let b = parser.push_block(block_t::event_block(event).within_unique_ptr());
-        parser.eval(
-            &buffer.to_ffi(),
-            &io_chain_t::new().within_unique_ptr(),
-            &SharedPtr::null(),
-            block_type_t::top,
-        );
-        parser.pop_block(b);
-        parser.set_last_statuses(prev_statuses);
+        // let parser = parser.pin();
+        // if false {
+        //     let b = parser.push_block(event_block(Box::new(event)).within_unique_ptr());
+        //     parser.eval(
+        //         &buffer.to_ffi(),
+        //         &io_chain_t::new().within_unique_ptr(),
+        //         &SharedPtr::null(),
+        //         block_type_t::top,
+        //     );
+        //     parser.pop_block(b);
+        //     parser.set_last_statuses(prev_statuses);
+        // }
 
         handler.fired.store(true, Ordering::Relaxed);
         fired_one_shot |= handler.is_one_shot();
@@ -517,7 +726,7 @@ fn fire_internal(parser: &mut parser_t, event: &Event) {
 }
 
 /// Handle all pending signal events.
-pub fn fire_delayed(parser: &mut parser_t) {
+pub fn event_fire_delayed(parser: &mut parser_t) {
     let ld = parser.get_libdata_pod();
     // Do not invoke new event handlers from within event handlers.
     if ld.is_event != 0 {
@@ -541,7 +750,7 @@ pub fn fire_delayed(parser: &mut parser_t) {
             // are $COLUMNS and $LINES.
             // Do that now.
             if i32::try_from(sig).expect("signal should be < 2^31") == SIGWINCH {
-                termsize_container_t::ffi_updating(parser.pin());
+                let _ = termsize_container_t::ffi_updating(parser.pin());
             }
             let e = Arc::new(Event {
                 desc: EventDescription {
@@ -566,14 +775,19 @@ pub fn fire_delayed(parser: &mut parser_t) {
     }
 }
 
+pub fn event_fire_delayed_ffi(parser: Pin<&mut parser_t>) {
+    event_fire_delayed(parser.unpin())
+}
+
+/// Enqueue a signal event. Invoked from a signal handler.
 pub fn enqueue_signal(signal: usize) {
     // Beware, we are in a signal handler
     s_pending_signals.mark(signal);
 }
 
-pub fn fire(parser: &mut parser_t, event: Event) {
+pub fn event_fire(parser: &mut parser_t, event: Event) {
     // Fire events triggered by signals.
-    fire_delayed(parser);
+    event_fire_delayed(parser);
 
     if event.is_blocked(parser) {
         s_blocked_events
@@ -585,6 +799,10 @@ pub fn fire(parser: &mut parser_t, event: Event) {
     }
 }
 
+fn event_fire_ffi(parser: Pin<&mut parser_t>, event: &Event) {
+    event_fire(parser.unpin(), event.clone())
+}
+
 #[widestrs]
 const EVENT_FILTER_NAMES: &[&wstr] = &[
     "signal"L,
@@ -596,6 +814,7 @@ const EVENT_FILTER_NAMES: &[&wstr] = &[
     "generic"L,
 ];
 
+/// Print all events. If type_filter is not empty, only output events with that type.
 #[widestrs]
 pub fn event_print(streams: &mut io_streams_t, type_filter: &wstr) {
     let mut tmp = s_event_handlers
@@ -645,8 +864,14 @@ pub fn event_print(streams: &mut io_streams_t, type_filter: &wstr) {
     }
 }
 
+fn event_print_ffi(streams: Pin<&mut ffi::io_streams_t>, type_filter: &CxxWString) {
+    let mut streams = io_streams_t::new(streams);
+    event_print(&mut streams, &type_filter.from_ffi());
+}
+
+/// Fire a generic event with the specified name.
 pub fn event_fire_generic(parser: &mut parser_t, name: WString, arguments: Vec<WString>) {
-    fire(
+    event_fire(
         parser,
         Event {
             desc: EventDescription {
@@ -656,3 +881,15 @@ pub fn event_fire_generic(parser: &mut parser_t, name: WString, arguments: Vec<W
         },
     )
 }
+
+fn event_fire_generic_ffi(
+    parser: Pin<&mut parser_t>,
+    name: &CxxWString,
+    arguments: &CxxVector<wcharz_t>,
+) {
+    event_fire_generic(
+        parser.unpin(),
+        name.from_ffi(),
+        arguments.iter().map(WString::from).collect(),
+    );
+}
diff --git a/fish-rust/src/ffi.rs b/fish-rust/src/ffi.rs
index ccadd8d32..3760bc800 100644
--- a/fish-rust/src/ffi.rs
+++ b/fish-rust/src/ffi.rs
@@ -81,6 +81,7 @@ include_cpp! {
     generate!("block_type_t")
     generate!("statuses_t")
     generate!("io_chain_t")
+    generate!("event_block")
 
     generate!("termsize_container_t")
 }
@@ -96,8 +97,8 @@ impl parser_t {
         unsafe { slice::from_raw_parts(ffi_jobs.jobs, ffi_jobs.count) }
     }
 
-    pub fn global_event_blocks(&self) -> &event_blockage_list_t {
-        &parser_t::ffi_global_event_blocks(self.pin())
+    pub fn global_event_blocks(&mut self) -> *mut event_blockage_list_t {
+        self.pin().ffi_global_event_blocks()
     }
 
     pub fn job_get_from_pid(&self, pid: pid_t) -> Option<&job_t> {
@@ -105,15 +106,15 @@ impl parser_t {
         unsafe { job.as_ref() }
     }
 
-    pub fn get_libdata_pod(&self) -> &mut library_data_pod {
-        let libdata = self.ffi_libdata_pod();
+    pub fn get_libdata_pod(&mut self) -> &mut library_data_pod {
+        let libdata = self.pin().ffi_libdata_pod();
 
         unsafe { &mut *libdata }
     }
 }
 
 impl block_t {
-    pub fn event_blocks(&self) -> &event_blockage_list_t {
+    pub fn event_blocks(&self) -> *const event_blockage_list_t {
         let blocks = self.ffi_event_blocks();
         unsafe { &*blocks }
     }
diff --git a/src/builtins/function.cpp b/src/builtins/function.cpp
index 1eca5b2ef..44a5caa12 100644
--- a/src/builtins/function.cpp
+++ b/src/builtins/function.cpp
@@ -101,7 +101,10 @@ static int parse_cmd_opts(function_cmd_opts_t &opts, int *optind,  //!OCLINT(hig
                     streams.err.append_format(_(L"%ls: Unknown signal '%ls'"), cmd, w.woptarg);
                     return STATUS_INVALID_ARGS;
                 }
-                opts.events.push_back(event_description_t::signal(sig));
+                event_description_t event_desc;
+                event_desc.typ = event_type_t::signal;
+                event_desc.signal = sig;
+                opts.events.push_back(std::move(event_desc));
                 break;
             }
             case 'v': {
@@ -110,16 +113,23 @@ static int parse_cmd_opts(function_cmd_opts_t &opts, int *optind,  //!OCLINT(hig
                     return STATUS_INVALID_ARGS;
                 }
 
-                opts.events.push_back(event_description_t::variable(w.woptarg));
+                event_description_t event_desc;
+                event_desc.typ = event_type_t::variable;
+                event_desc.str_param1 = std::make_unique<wcstring>(w.woptarg);
+                opts.events.push_back(std::move(event_desc));
                 break;
             }
             case 'e': {
-                opts.events.push_back(event_description_t::generic(w.woptarg));
+                event_description_t event_desc;
+                event_desc.typ = event_type_t::generic;
+                event_desc.str_param1 = std::make_unique<wcstring>(w.woptarg);
+                opts.events.push_back(std::move(event_desc));
                 break;
             }
             case 'j':
             case 'p': {
-                event_description_t e(event_type_t::any);
+                event_description_t e;
+                e.typ = event_type_t::any;
 
                 if ((opt == 'j') && (wcscasecmp(w.woptarg, L"caller") == 0)) {
                     internal_job_id_t caller_id =
@@ -129,11 +139,11 @@ static int parse_cmd_opts(function_cmd_opts_t &opts, int *optind,  //!OCLINT(hig
                             _(L"%ls: calling job for event handler not found"), cmd);
                         return STATUS_INVALID_ARGS;
                     }
-                    e.type = event_type_t::caller_exit;
-                    e.param1.caller_id = caller_id;
+                    e.typ = event_type_t::caller_exit;
+                    e.caller_id = caller_id;
                 } else if ((opt == 'p') && (wcscasecmp(w.woptarg, L"%self") == 0)) {
-                    e.type = event_type_t::process_exit;
-                    e.param1.pid = getpid();
+                    e.typ = event_type_t::process_exit;
+                    e.pid = getpid();
                 } else {
                     pid_t pid = fish_wcstoi(w.woptarg);
                     if (errno || pid < 0) {
@@ -142,14 +152,15 @@ static int parse_cmd_opts(function_cmd_opts_t &opts, int *optind,  //!OCLINT(hig
                         return STATUS_INVALID_ARGS;
                     }
                     if (opt == 'p') {
-                        e.type = event_type_t::process_exit;
-                        e.param1.pid = pid;
+                        e.typ = event_type_t::process_exit;
+                        e.pid = pid;
                     } else {
-                        e.type = event_type_t::job_exit;
-                        e.param1.jobspec = {pid, job_id_for_pid(pid, parser)};
+                        e.typ = event_type_t::job_exit;
+                        e.pid = pid;
+                        e.internal_job_id = job_id_for_pid(pid, parser);
                     }
                 }
-                opts.events.push_back(e);
+                opts.events.push_back(std::move(e));
                 break;
             }
             case 'a': {
@@ -294,25 +305,25 @@ int builtin_function(parser_t &parser, io_streams_t &streams, const wcstring_lis
 
     // Add any event handlers.
     for (const event_description_t &ed : opts.events) {
-        event_add_handler(std::make_shared<event_handler_t>(ed, function_name));
+        event_add_handler(ed, function_name);
     }
 
     // If there is an --on-process-exit or --on-job-exit event handler for some pid, and that
     // process has already exited, run it immediately (#7210).
     for (const event_description_t &ed : opts.events) {
-        if (ed.type == event_type_t::process_exit) {
-            pid_t pid = ed.param1.pid;
+        if (ed.typ == event_type_t::process_exit) {
+            pid_t pid = ed.pid;
             if (pid == EVENT_ANY_PID) continue;
             wait_handle_ref_t wh = parser.get_wait_handles().get_by_pid(pid);
             if (wh && wh->completed) {
-                event_fire(parser, event_t::process_exit(pid, wh->status));
+                event_fire(parser, *new_event_process_exit(pid, wh->status));
             }
-        } else if (ed.type == event_type_t::job_exit) {
-            pid_t pid = ed.param1.jobspec.pid;
+        } else if (ed.typ == event_type_t::job_exit) {
+            pid_t pid = ed.pid;
             if (pid == EVENT_ANY_PID) continue;
             wait_handle_ref_t wh = parser.get_wait_handles().get_by_pid(pid);
             if (wh && wh->completed) {
-                event_fire(parser, event_t::job_exit(pid, wh->internal_job_id));
+                event_fire(parser, *new_event_job_exit(pid, wh->internal_job_id));
             }
         }
     }
diff --git a/src/builtins/set.cpp b/src/builtins/set.cpp
index 9417b884d..d6d3b6e94 100644
--- a/src/builtins/set.cpp
+++ b/src/builtins/set.cpp
@@ -652,7 +652,7 @@ static int builtin_set_erase(const wchar_t *cmd, set_cmd_opts_t &opts, int argc,
                     handle_env_return(retval, cmd, split->varname, streams);
                 }
                 if (retval == ENV_OK) {
-                    event_fire(parser, event_t::variable_erase(split->varname));
+                    event_fire(parser, *new_event_variable_erase(split->varname));
                 }
             } else {  // remove just the specified indexes of the var
                 if (!split->var) return STATUS_CMD_ERROR;
diff --git a/src/env.cpp b/src/env.cpp
index 38d9fc5f6..0ff8490b7 100644
--- a/src/env.cpp
+++ b/src/env.cpp
@@ -1300,7 +1300,7 @@ mod_result_t env_stack_impl_t::remove(const wcstring &key, int mode) {
     return result;
 }
 
-std::vector<event_t> env_stack_t::universal_sync(bool always) {
+std::vector<rust::Box<Event>> env_stack_t::universal_sync(bool always) {
     if (s_uvar_scope_is_global) return {};
     if (!always && !s_uvars_locally_modified) return {};
     s_uvars_locally_modified = false;
@@ -1311,11 +1311,11 @@ std::vector<event_t> env_stack_t::universal_sync(bool always) {
         universal_notifier_t::default_notifier().post_notification();
     }
     // React internally to changes to special variables like LANG, and populate on-variable events.
-    std::vector<event_t> result;
+    std::vector<rust::Box<Event>> result;
     for (const callback_data_t &cb : callbacks) {
         env_dispatch_var_change(cb.key, *this);
-        event_t evt =
-            cb.is_erase() ? event_t::variable_erase(cb.key) : event_t::variable_set(cb.key);
+        auto evt =
+            cb.is_erase() ? new_event_variable_erase(cb.key) : new_event_variable_set(cb.key);
         result.push_back(std::move(evt));
     }
     return result;
diff --git a/src/env.h b/src/env.h
index 972846efd..eea1558a9 100644
--- a/src/env.h
+++ b/src/env.h
@@ -13,6 +13,7 @@
 #include <vector>
 
 #include "common.h"
+#include "cxx.h"
 #include "maybe.h"
 
 class owning_null_terminated_array_t;
@@ -20,7 +21,7 @@ class owning_null_terminated_array_t;
 extern size_t read_byte_limit;
 extern bool curses_initialized;
 
-struct event_t;
+struct Event;
 
 // Flags that may be passed as the 'mode' in env_stack_t::set() / environment_t::get().
 enum : uint16_t {
@@ -287,7 +288,7 @@ class env_stack_t final : public environment_t {
     /// If \p always is set, perform synchronization even if there's no pending changes from this
     /// instance (that is, look for changes from other fish instances).
     /// \return a list of events for changed variables.
-    std::vector<event_t> universal_sync(bool always);
+    std::vector<rust::Box<Event>> universal_sync(bool always);
 
     // Compatibility hack; access the "environment stack" from back when there was just one.
     static const std::shared_ptr<env_stack_t> &principal_ref();
diff --git a/src/event.cpp b/src/event.cpp
index da8c562c4..247479a52 100644
--- a/src/event.cpp
+++ b/src/event.cpp
@@ -92,9 +92,6 @@ static owning_lock<event_handler_list_t> s_event_handlers;
 /// This is inspected by a signal handler. We assume no values in here overflow.
 static std::array<relaxed_atomic_t<uint32_t>, NSIG> s_observed_signals;
 
-/// List of events that have been sent but have not yet been delivered because they are blocked.
-static std::vector<std::shared_ptr<const event_t>> s_blocked_events{};
-
 static inline void inc_signal_observed(int sig) {
     if (0 <= sig && sig < NSIG) {
         s_observed_signals[sig]++;
@@ -107,125 +104,10 @@ static inline void dec_signal_observed(int sig) {
     }
 }
 
-bool event_is_signal_observed(int sig) {
-    // We are in a signal handler!
-    uint32_t count = 0;
-    if (0 <= sig && sig < NSIG) {
-        count = s_observed_signals[sig];
-    }
-    return count > 0;
-}
-
-/// \return true if a handler is "one shot": it fires at most once.
-static bool handler_is_one_shot(const event_handler_t &handler) {
-    switch (handler.desc.type) {
-        case event_type_t::process_exit:
-            return handler.desc.param1.pid != EVENT_ANY_PID;
-        case event_type_t::job_exit:
-            return handler.desc.param1.jobspec.pid != EVENT_ANY_PID;
-        case event_type_t::caller_exit:
-            return true;
-        case event_type_t::signal:
-        case event_type_t::variable:
-        case event_type_t::generic:
-        case event_type_t::any:
-            return false;
-    }
-    DIE("Unreachable");
-}
-
-/// Tests if one event instance matches the definition of an event class.
-/// In case of a match, \p only_once indicates that the event cannot match again by nature.
-static bool handler_matches(const event_handler_t &handler, const event_t &instance) {
-    if (handler.desc.type == event_type_t::any) return true;
-    if (handler.desc.type != instance.desc.type) return false;
-
-    switch (handler.desc.type) {
-        case event_type_t::signal: {
-            return handler.desc.param1.signal == instance.desc.param1.signal;
-        }
-        case event_type_t::variable: {
-            return instance.desc.str_param1 == handler.desc.str_param1;
-        }
-        case event_type_t::process_exit: {
-            if (handler.desc.param1.pid == EVENT_ANY_PID) return true;
-            return handler.desc.param1.pid == instance.desc.param1.pid;
-        }
-        case event_type_t::job_exit: {
-            const auto &jobspec = handler.desc.param1.jobspec;
-            if (jobspec.pid == EVENT_ANY_PID) return true;
-            return jobspec.internal_job_id == instance.desc.param1.jobspec.internal_job_id;
-        }
-        case event_type_t::caller_exit: {
-            return handler.desc.param1.caller_id == instance.desc.param1.caller_id;
-        }
-        case event_type_t::generic: {
-            return handler.desc.str_param1 == instance.desc.str_param1;
-        }
-        case event_type_t::any:
-        default: {
-            DIE("unexpected classv.type");
-            return false;
-        }
-    }
-}
-
-/// Test if specified event is blocked.
-static bool event_is_blocked(parser_t &parser, const event_t &e) {
-    (void)e;
-    const block_t *block;
-    size_t idx = 0;
-    while ((block = parser.block_at_index(idx++))) {
-        if (event_block_list_blocks_type(block->event_blocks)) return true;
-    }
-    return event_block_list_blocks_type(parser.global_event_blocks);
-}
-
-wcstring event_get_desc(const parser_t &parser, const event_t &evt) {
-    const event_description_t &ed = evt.desc;
-    switch (ed.type) {
-        case event_type_t::signal: {
-            return format_string(_(L"signal handler for %ls (%ls)"), sig2wcs(ed.param1.signal),
-                                 signal_get_desc(ed.param1.signal));
-        }
-
-        case event_type_t::variable: {
-            return format_string(_(L"handler for variable '%ls'"), ed.str_param1.c_str());
-        }
-
-        case event_type_t::process_exit: {
-            return format_string(_(L"exit handler for process %d"), ed.param1.pid);
-        }
-
-        case event_type_t::job_exit: {
-            const auto &jobspec = ed.param1.jobspec;
-            if (const job_t *j = parser.job_get_from_pid(jobspec.pid)) {
-                return format_string(_(L"exit handler for job %d, '%ls'"), j->job_id(),
-                                     j->command_wcstr());
-            } else {
-                return format_string(_(L"exit handler for job with pid %d"), jobspec.pid);
-            }
-        }
-
-        case event_type_t::caller_exit: {
-            return _(L"exit handler for command substitution caller");
-        }
-
-        case event_type_t::generic: {
-            return format_string(_(L"handler for generic event '%ls'"), ed.str_param1.c_str());
-        }
-        case event_type_t::any: {
-            DIE("Unreachable");
-        }
-        default:
-            DIE("Unknown event type");
-    }
-}
-
-void event_add_handler(std::shared_ptr<event_handler_t> eh) {
-    if (eh->desc.type == event_type_t::signal) {
-        signal_handle(eh->desc.param1.signal);
-        inc_signal_observed(eh->desc.param1.signal);
+void event_add_handler(std::shared_ptr<rust::Box<EventHandler>> eh) {
+    if ((*eh)->desc().typ == event_type_t::signal) {
+        signal_handle((*eh)->desc().signal);
+        inc_signal_observed((*eh)->desc().signal);
     }
 
     s_event_handlers.acquire()->push_back(std::move(eh));
@@ -238,11 +120,11 @@ static void remove_handlers_if(const T &func) {
     auto handlers = s_event_handlers.acquire();
     auto iter = handlers->begin();
     while (iter != handlers->end()) {
-        event_handler_t *handler = iter->get();
+        rust::Box<EventHandler> *handler = iter->get();
         if (func(*handler)) {
-            handler->removed = true;
-            if (handler->desc.type == event_type_t::signal) {
-                dec_signal_observed(handler->desc.param1.signal);
+            (*handler)->set_removed();
+            if ((*handler)->desc().typ == event_type_t::signal) {
+                dec_signal_observed((*handler)->desc().signal);
             }
             iter = handlers->erase(iter);
         } else {
@@ -253,323 +135,26 @@ static void remove_handlers_if(const T &func) {
 
 void event_remove_function_handlers(const wcstring &name) {
     remove_handlers_if(
-        [&](const event_handler_t &handler) { return handler.function_name == name; });
+        [&](const rust::Box<EventHandler> &handler) { return *handler->function_name() == name; });
 }
 
 event_handler_list_t event_get_function_handlers(const wcstring &name) {
     auto handlers = s_event_handlers.acquire();
     event_handler_list_t result;
-    for (const shared_ptr<event_handler_t> &eh : *handlers) {
-        if (eh->function_name == name) {
+    for (const shared_ptr<rust::Box<EventHandler>> &eh : *handlers) {
+        if (*(*eh)->function_name() == name) {
             result.push_back(eh);
         }
     }
     return result;
 }
 
-/// Perform the specified event. Since almost all event firings will not be matched by even a single
-/// event handler, we make sure to optimize the 'no matches' path. This means that nothing is
-/// allocated/initialized unless needed.
-static void event_fire_internal(parser_t &parser, const event_t &event) {
-    auto &ld = parser.libdata();
-    assert(ld.pod.is_event >= 0 && "is_event should not be negative");
-    scoped_push<decltype(ld.pod.is_event)> inc_event{&ld.pod.is_event, ld.pod.is_event + 1};
-
-    // Suppress fish_trace during events.
-    scoped_push<bool> suppress_trace{&ld.pod.suppress_fish_trace, true};
-
-    // Capture the event handlers that match this event.
-    std::vector<std::shared_ptr<event_handler_t>> fire;
-    {
-        auto event_handlers = s_event_handlers.acquire();
-        for (const auto &handler : *event_handlers) {
-            if (handler_matches(*handler, event)) {
-                fire.push_back(handler);
-            }
-        }
-    }
-
-    // Iterate over our list of matching events. Fire the ones that are still present.
-    bool fired_one_shot = false;
-    for (const auto &handler : fire) {
-        // A previous handlers may have erased this one.
-        if (handler->removed) continue;
-
-        // Construct a buffer to evaluate, starting with the function name and then all the
-        // arguments.
-        wcstring buffer = handler->function_name;
-        for (const wcstring &arg : event.arguments) {
-            buffer.push_back(L' ');
-            buffer.append(escape_string(arg));
-        }
-
-        // Event handlers are not part of the main flow of code, so they are marked as
-        // non-interactive.
-        scoped_push<bool> interactive{&ld.pod.is_interactive, false};
-        auto prev_statuses = parser.get_last_statuses();
-
-        FLOGF(event, L"Firing event '%ls' to handler '%ls'", event.desc.str_param1.c_str(),
-              handler->function_name.c_str());
-        block_t *b = parser.push_block(block_t::event_block(event));
-        parser.eval(buffer, io_chain_t());
-        parser.pop_block(b);
-        parser.set_last_statuses(std::move(prev_statuses));
-
-        handler->fired = true;
-        fired_one_shot |= handler_is_one_shot(*handler);
-    }
-
-    // Remove any fired one-shot handlers.
-    if (fired_one_shot) {
-        remove_handlers_if([](const event_handler_t &handler) {
-            return handler.fired && handler_is_one_shot(handler);
-        });
-    }
-}
-
-/// Handle all pending signal events.
-void event_fire_delayed(parser_t &parser) {
-    auto &ld = parser.libdata();
-    // Do not invoke new event handlers from within event handlers.
-    if (ld.pod.is_event) return;
-    // Do not invoke new event handlers if we are unwinding (#6649).
-    if (signal_check_cancel()) return;
-
-    std::vector<shared_ptr<const event_t>> to_send;
-    to_send.swap(s_blocked_events);
-    assert(s_blocked_events.empty());
-
-    // Append all signal events to to_send.
-    auto signals = s_pending_signals.acquire_pending();
-    if (signals.any()) {
-        for (uint32_t sig = 0; sig < signals.size(); sig++) {
-            if (signals.test(sig)) {
-                // HACK: The only variables we change in response to a *signal*
-                // are $COLUMNS and $LINES.
-                // Do that now.
-                if (sig == SIGWINCH) {
-                    (void)termsize_container_t::shared().updating(parser);
-                }
-                auto e = std::make_shared<event_t>(event_type_t::signal);
-                e->desc.param1.signal = sig;
-                e->arguments.push_back(sig2wcs(sig));
-                to_send.push_back(std::move(e));
-            }
-        }
-    }
-
-    // Fire or re-block all events.
-    for (const auto &evt : to_send) {
-        if (event_is_blocked(parser, *evt)) {
-            s_blocked_events.push_back(evt);
-        } else {
-            event_fire_internal(parser, *evt);
-        }
-    }
-}
-
-void event_enqueue_signal(int signal) {
-    // Beware, we are in a signal handler
-    s_pending_signals.mark(signal);
-}
-
-void event_fire(parser_t &parser, const event_t &event) {
-    // Fire events triggered by signals.
-    event_fire_delayed(parser);
-
-    if (event_is_blocked(parser, event)) {
-        s_blocked_events.push_back(std::make_shared<event_t>(event));
-    } else {
-        event_fire_internal(parser, event);
-    }
-}
-
-static const wchar_t *event_name_for_type(event_type_t type) {
-    switch (type) {
-        case event_type_t::any:
-            return L"any";
-        case event_type_t::signal:
-            return L"signal";
-        case event_type_t::variable:
-            return L"variable";
-        case event_type_t::process_exit:
-            return L"process-exit";
-        case event_type_t::job_exit:
-            return L"job-exit";
-        case event_type_t::caller_exit:
-            return L"caller-exit";
-        case event_type_t::generic:
-            return L"generic";
-    }
-    return L"";
-}
-
 const wchar_t *const event_filter_names[] = {L"signal",       L"variable", L"exit",
                                              L"process-exit", L"job-exit", L"caller-exit",
                                              L"generic",      nullptr};
 
-static bool filter_matches_event(const wcstring &filter, event_type_t type) {
-    if (filter.empty()) return true;
-    switch (type) {
-        case event_type_t::any:
-            return false;
-        case event_type_t::signal:
-            return filter == L"signal";
-        case event_type_t::variable:
-            return filter == L"variable";
-        case event_type_t::process_exit:
-            return filter == L"process-exit" || filter == L"exit";
-        case event_type_t::job_exit:
-            return filter == L"job-exit" || filter == L"exit";
-        case event_type_t::caller_exit:
-            return filter == L"caller-exit" || filter == L"exit";
-        case event_type_t::generic:
-            return filter == L"generic";
-    }
-    DIE("Unreachable");
-}
-
-void event_print(io_streams_t &streams, const wcstring &type_filter) {
-    event_handler_list_t tmp = *s_event_handlers.acquire();
-    std::sort(tmp.begin(), tmp.end(),
-              [](const shared_ptr<event_handler_t> &e1, const shared_ptr<event_handler_t> &e2) {
-                  const event_description_t &d1 = e1->desc;
-                  const event_description_t &d2 = e2->desc;
-                  if (d1.type != d2.type) {
-                      return d1.type < d2.type;
-                  }
-                  switch (d1.type) {
-                      case event_type_t::signal:
-                          return d1.param1.signal < d2.param1.signal;
-                      case event_type_t::process_exit:
-                          return d1.param1.pid < d2.param1.pid;
-                      case event_type_t::job_exit:
-                          return d1.param1.jobspec.pid < d2.param1.jobspec.pid;
-                      case event_type_t::caller_exit:
-                          return d1.param1.caller_id < d2.param1.caller_id;
-                      case event_type_t::variable:
-                      case event_type_t::any:
-                      case event_type_t::generic:
-                          return d1.str_param1 < d2.str_param1;
-                  }
-                  DIE("Unreachable");
-              });
-
-    maybe_t<event_type_t> last_type{};
-    for (const shared_ptr<event_handler_t> &evt : tmp) {
-        // If we have a filter, skip events that don't match.
-        if (!filter_matches_event(type_filter, evt->desc.type)) {
-            continue;
-        }
-
-        if (!last_type || *last_type != evt->desc.type) {
-            if (last_type) streams.out.append(L"\n");
-            last_type = evt->desc.type;
-            streams.out.append_format(L"Event %ls\n", event_name_for_type(*last_type));
-        }
-        switch (evt->desc.type) {
-            case event_type_t::signal:
-                streams.out.append_format(L"%ls %ls\n", sig2wcs(evt->desc.param1.signal),
-                                          evt->function_name.c_str());
-                break;
-            case event_type_t::process_exit:
-            case event_type_t::job_exit:
-                break;
-            case event_type_t::caller_exit:
-                streams.out.append_format(L"caller-exit %ls\n", evt->function_name.c_str());
-                break;
-            case event_type_t::variable:
-            case event_type_t::generic:
-                streams.out.append_format(L"%ls %ls\n", evt->desc.str_param1.c_str(),
-                                          evt->function_name.c_str());
-                break;
-            case event_type_t::any:
-                DIE("Unreachable");
-            default:
-                streams.out.append_format(L"%ls\n", evt->function_name.c_str());
-                break;
-        }
-    }
-}
-
-void event_fire_generic(parser_t &parser, wcstring name, const wcharz_t *argv, int argc) {
-    wcstring_list_t args_vec{};
-    for (int i = 0; i < argc; i++) {
-        args_vec.push_back(argv[i]);
-    }
-    event_fire_generic(parser, std::move(name), std::move(args_vec));
-}
-
-void event_fire_generic(parser_t &parser, wcstring name, wcstring_list_t args) {
-    event_t ev(event_type_t::generic);
-    ev.desc.str_param1 = std::move(name);
-    ev.arguments = std::move(args);
-    event_fire(parser, ev);
-}
-
-event_description_t event_description_t::signal(int sig) {
-    event_description_t event(event_type_t::signal);
-    event.param1.signal = sig;
-    return event;
-}
-
-event_description_t event_description_t::variable(wcstring str) {
-    event_description_t event(event_type_t::variable);
-    event.str_param1 = std::move(str);
-    return event;
-}
-
-event_description_t event_description_t::generic(wcstring str) {
-    event_description_t event(event_type_t::generic);
-    event.str_param1 = std::move(str);
-    return event;
-}
-
-// static
-event_t event_t::variable_erase(wcstring name) {
-    event_t evt{event_type_t::variable};
-    evt.arguments = {L"VARIABLE", L"ERASE", name};
-    evt.desc.str_param1 = std::move(name);
-    return evt;
-}
-
-// static
-event_t event_t::variable_set(wcstring name) {
-    event_t evt{event_type_t::variable};
-    evt.arguments = {L"VARIABLE", L"SET", name};
-    evt.desc.str_param1 = std::move(name);
-    return evt;
-}
-
-// static
-event_t event_t::process_exit(pid_t pid, int status) {
-    event_t evt{event_type_t::process_exit};
-    evt.desc.param1.pid = pid;
-    evt.arguments.reserve(3);
-    evt.arguments.push_back(L"PROCESS_EXIT");
-    evt.arguments.push_back(to_string(pid));
-    evt.arguments.push_back(to_string(status));
-    return evt;
-}
-
-// static
-event_t event_t::job_exit(pid_t pgid, internal_job_id_t jid) {
-    event_t evt{event_type_t::job_exit};
-    evt.desc.param1.jobspec = {pgid, jid};
-    evt.arguments.reserve(3);
-    evt.arguments.push_back(L"JOB_EXIT");
-    evt.arguments.push_back(to_string(pgid));
-    evt.arguments.push_back(L"0");  // historical
-    return evt;
-}
-
-// static
-event_t event_t::caller_exit(uint64_t internal_job_id, int job_id) {
-    event_t evt{event_type_t::caller_exit};
-    evt.desc.param1.caller_id = internal_job_id;
-    evt.arguments.reserve(3);
-    evt.arguments.push_back(L"JOB_EXIT");
-    evt.arguments.push_back(to_string(job_id));
-    evt.arguments.push_back(L"0");  // historical
-    return evt;
+void event_fire_generic(parser_t &parser, const wcstring &name, wcstring_list_t args) {
+    std::vector<wcharz_t> ffi_args;
+    for (const auto &arg : args) ffi_args.push_back(arg.c_str());
+    event_fire_generic_ffi(parser, name, ffi_args);
 }
diff --git a/src/event.h b/src/event.h
index c7b2380c2..44d4da91d 100644
--- a/src/event.h
+++ b/src/event.h
@@ -1,8 +1,3 @@
-// Functions for handling event triggers
-//
-// Because most of these functions can be called by signal handler, it is important to make it well
-// defined when these functions produce output or perform memory allocations, since such functions
-// may not be safely called by signal handlers.
 #ifndef FISH_EVENT_H
 #define FISH_EVENT_H
 
@@ -17,156 +12,38 @@
 #include "global_safety.h"
 #include "wutil.h"
 
+#if INCLUDE_RUST_HEADERS
+
+#include "event.rs.h"
+
+#else
+
+// Hacks to allow us to compile without Rust headers.
+
+struct Event;
+struct EventHandler;
+
+#endif
+
 struct io_streams_t;
 
 /// The process id that is used to match any process id.
 #define EVENT_ANY_PID 0
 
-/// Enumeration of event types.
-enum class event_type_t {
-    /// Matches any event type (Not always any event, as the function name may limit the choice as
-    /// well.
-    any,
-    /// An event triggered by a signal.
-    signal,
-    /// An event triggered by a variable update.
-    variable,
-    /// An event triggered by a process exit.
-    process_exit,
-    /// An event triggered by a job exit.
-    job_exit,
-    /// An event triggered by a job exit, triggering the 'caller'-style events only.
-    caller_exit,
-    /// A generic event.
-    generic,
-};
-
 /// Null-terminated list of valid event filter names.
 /// These are what are valid to pass to 'functions --handlers-type'
 extern const wchar_t *const event_filter_names[];
 
-/// Properties of an event.
-struct event_description_t {
-    /// Helper type for on-job-exit events.
-    struct job_spec_t {
-        // pid requested by the event, or ANY_PID for all.
-        pid_t pid;
-
-        // internal_job_id of the job to match.
-        // If this is 0, we match either all jobs (pid == ANY_PID) or no jobs (otherwise).
-        uint64_t internal_job_id;
-    };
-
-    /// The event type.
-    event_type_t type;
-
-    /// The type-specific parameter. The int types are one of the following:
-    ///
-    /// signal: Signal number for signal-type events.Use EVENT_ANY_SIGNAL to match any signal
-    /// pid: Process id for process-type events. Use EVENT_ANY_PID to match any pid.
-    /// jobspec: Info for on-job-exit events.
-    /// caller_id: Internal job id for caller_exit type events
-    union {
-        int signal;
-        pid_t pid;
-        job_spec_t jobspec;
-        uint64_t caller_id;
-    } param1{};
-
-    /// The string types are one of the following:
-    ///
-    /// variable: Variable name for variable-type events.
-    /// param: The parameter describing this generic event.
-    wcstring str_param1{};
-
-    explicit event_description_t(event_type_t t) : type(t) {}
-    static event_description_t signal(int sig);
-    static event_description_t variable(wcstring str);
-    static event_description_t generic(wcstring str);
-};
-
-/// Represents a handler for an event.
-struct event_handler_t {
-    /// Properties of the event to match.
-    const event_description_t desc;
-
-    /// Name of the function to invoke.
-    const wcstring function_name{};
-
-    /// A flag set when an event handler is removed from the global list.
-    /// Once set, this is never cleared.
-    relaxed_atomic_bool_t removed{false};
-
-    /// A flag set when an event handler is first fired.
-    relaxed_atomic_bool_t fired{false};
-
-    explicit event_handler_t(event_type_t t) : desc(std::move(t)) {}
-
-    event_handler_t(event_description_t d, wcstring name)
-        : desc(std::move(d)), function_name(std::move(name)) {}
-};
-using event_handler_list_t = std::vector<std::shared_ptr<event_handler_t>>;
-
-/// Represents a event that is fired, or capable of being fired.
-struct event_t {
-    /// Properties of the event.
-    event_description_t desc;
-
-    /// Arguments to any handler.
-    wcstring_list_t arguments{};
-
-    explicit event_t(event_type_t t) : desc(t) {}
-
-    /// Create an event_type_t::variable event with the args for erasing a variable.
-    static event_t variable_erase(wcstring name);
-    /// Create an event_type_t::variable event with the args for setting a variable.
-    static event_t variable_set(wcstring name);
-
-    /// Create a PROCESS_EXIT event.
-    static event_t process_exit(pid_t pid, int status);
-
-    /// Create a JOB_EXIT event. The pgid should be positive.
-    /// The reported status is always 0 for historical reasons.
-    static event_t job_exit(pid_t pgid, internal_job_id_t jid);
-
-    /// Create a caller_exit event.
-    static event_t caller_exit(uint64_t internal_job_id, int job_id);
-};
+using event_handler_list_t = std::vector<std::shared_ptr<rust::Box<EventHandler>>>;
 
 class parser_t;
 
-/// Add an event handler.
-void event_add_handler(std::shared_ptr<event_handler_t> eh);
-
 /// Remove all events for the given function name.
 void event_remove_function_handlers(const wcstring &name);
 
 /// Return all event handlers for the given function.
 event_handler_list_t event_get_function_handlers(const wcstring &name);
 
-/// Returns whether an event listener is registered for the given signal. This is safe to call from
-/// a signal handler.
-bool event_is_signal_observed(int signal);
-
-/// Fire the specified event \p event, executing it on \p parser.
-void event_fire(parser_t &parser, const event_t &event);
-
-/// Fire all delayed events attached to the given parser.
-void event_fire_delayed(parser_t &parser);
-
-/// Enqueue a signal event. Invoked from a signal handler.
-void event_enqueue_signal(int signal);
-
-/// Print all events. If type_filter is not empty, only output events with that type.
-void event_print(io_streams_t &streams, const wcstring &type_filter);
-
-/// Returns a string describing the specified event.
-wcstring event_get_desc(const parser_t &parser, const event_t &e);
-
-// FFI helper for event_fire_generic
-void event_fire_generic(parser_t &parser, wcstring name, const wcharz_t *argv, int argc);
-
-/// Fire a generic event with the specified name.
-void event_fire_generic(parser_t &parser, wcstring name, wcstring_list_t args = {});
+void event_fire_generic(parser_t &parser, const wcstring &name, wcstring_list_t args = {});
 
 #endif
diff --git a/src/fish.cpp b/src/fish.cpp
index b602bd8e1..eed026673 100644
--- a/src/fish.cpp
+++ b/src/fish.cpp
@@ -597,7 +597,7 @@ int main(int argc, char **argv) {
     }
 
     int exit_status = res ? STATUS_CMD_UNKNOWN : parser.get_last_status();
-    event_fire(parser, event_t::process_exit(getpid(), exit_status));
+    event_fire(parser, *new_event_process_exit(getpid(), exit_status));
 
     // Trigger any exit handlers.
     event_fire_generic(parser, L"fish_exit", {to_string(exit_status)});
diff --git a/src/function.cpp b/src/function.cpp
index 283d57233..c2606f408 100644
--- a/src/function.cpp
+++ b/src/function.cpp
@@ -288,7 +288,7 @@ wcstring function_properties_t::annotated_definition(const wcstring &name) const
     wcstring out;
     wcstring desc = this->localized_description();
     wcstring def = get_function_body_source(*this);
-    std::vector<std::shared_ptr<event_handler_t>> ev = event_get_function_handlers(name);
+    std::vector<std::shared_ptr<rust::Box<EventHandler>>> ev = event_get_function_handlers(name);
 
     out.append(L"function ");
 
@@ -315,22 +315,22 @@ wcstring function_properties_t::annotated_definition(const wcstring &name) const
     }
 
     for (const auto &next : ev) {
-        const event_description_t &d = next->desc;
-        switch (d.type) {
+        const event_description_t &d = (*next)->desc();
+        switch (d.typ) {
             case event_type_t::signal: {
-                append_format(out, L" --on-signal %ls", sig2wcs(d.param1.signal));
+                append_format(out, L" --on-signal %ls", sig2wcs(d.signal));
                 break;
             }
             case event_type_t::variable: {
-                append_format(out, L" --on-variable %ls", d.str_param1.c_str());
+                append_format(out, L" --on-variable %ls", d.str_param1->c_str());
                 break;
             }
             case event_type_t::process_exit: {
-                append_format(out, L" --on-process-exit %d", d.param1.pid);
+                append_format(out, L" --on-process-exit %d", d.pid);
                 break;
             }
             case event_type_t::job_exit: {
-                append_format(out, L" --on-job-exit %d", d.param1.jobspec.pid);
+                append_format(out, L" --on-job-exit %d", d.pid);
                 break;
             }
             case event_type_t::caller_exit: {
@@ -338,12 +338,12 @@ wcstring function_properties_t::annotated_definition(const wcstring &name) const
                 break;
             }
             case event_type_t::generic: {
-                append_format(out, L" --on-event %ls", d.str_param1.c_str());
+                append_format(out, L" --on-event %ls", d.str_param1->c_str());
                 break;
             }
             case event_type_t::any:
             default: {
-                DIE("unexpected next->type");
+                DIE("unexpected next->typ");
             }
         }
     }
diff --git a/src/parse_execution.cpp b/src/parse_execution.cpp
index 997164f7d..b099f3ba0 100644
--- a/src/parse_execution.cpp
+++ b/src/parse_execution.cpp
@@ -463,7 +463,7 @@ end_execution_reason_t parse_execution_context_t::run_for_statement(
     block_t *fb = parser->push_block(block_t::for_block());
 
     // We fire the same event over and over again, just construct it once.
-    event_t evt = event_t::variable_set(for_var_name);
+    auto evt = new_event_variable_set(for_var_name);
 
     // Now drive the for loop.
     for (const wcstring &val : arguments) {
@@ -475,7 +475,7 @@ end_execution_reason_t parse_execution_context_t::run_for_statement(
         retval = vars.set(for_var_name, ENV_DEFAULT | ENV_USER, {val});
         assert(retval == ENV_OK && "for loop variable should have been successfully set");
         (void)retval;
-        event_fire(*parser, evt);
+        event_fire(*parser, *evt);
 
         auto &ld = parser->libdata();
         ld.loop_status = loop_status_t::normals;
@@ -786,9 +786,8 @@ end_execution_reason_t parse_execution_context_t::handle_command_not_found(
         }
         auto prev_statuses = parser->get_last_statuses();
 
-        event_t event(event_type_t::generic);
-        event.desc.str_param1 = L"fish_command_not_found";
-        block_t *b = parser->push_block(block_t::event_block(event));
+        auto event = new_event_generic(L"fish_command_not_found");
+        block_t *b = parser->push_block(event_block(std::move(event)));
         parser->eval(buffer, io);
         parser->pop_block(b);
         parser->set_last_statuses(std::move(prev_statuses));
diff --git a/src/parser.cpp b/src/parser.cpp
index 651e68258..9890462ed 100644
--- a/src/parser.cpp
+++ b/src/parser.cpp
@@ -65,7 +65,7 @@ void parser_t::assert_can_execute() const { ASSERT_IS_MAIN_THREAD(); }
 int parser_t::set_var_and_fire(const wcstring &key, env_mode_flags_t mode, wcstring_list_t vals) {
     int res = vars().set(key, mode, std::move(vals));
     if (res == ENV_OK) {
-        event_fire(*this, event_t::variable_set(key));
+        event_fire(*this, *new_event_variable_set(key));
     }
     return res;
 }
@@ -80,7 +80,7 @@ void parser_t::sync_uvars_and_fire(bool always) {
     if (this->syncs_uvars_) {
         auto evts = this->vars().universal_sync(always);
         for (const auto &evt : evts) {
-            event_fire(*this, evt);
+            event_fire(*this, *evt);
         }
     }
 }
@@ -252,7 +252,7 @@ static void append_block_description_to_stack_trace(const parser_t &parser, cons
         }
         case block_type_t::event: {
             assert(b.event && "Should have an event");
-            wcstring description = event_get_desc(parser, *b.event);
+            wcstring description = *event_get_desc(parser, **b.event);
             append_format(trace, _(L"in event handler: %ls\n"), description.c_str());
             print_call_site = true;
             break;
@@ -491,7 +491,7 @@ job_t *parser_t::job_get_from_pid(pid_t pid) const {
 
 job_t *parser_t::ffi_job_get_from_pid(int pid) const { return job_get_from_pid(pid); }
 
-library_data_pod *parser_t::ffi_libdata_pod() const { return &library_data.pod; }
+library_data_pod *parser_t::ffi_libdata_pod() { return &library_data.pod; }
 
 profile_item_t *parser_t::create_profile_item() {
     if (g_profiling_active) {
@@ -745,9 +745,9 @@ wcstring block_t::description() const {
 
 block_t block_t::if_block() { return block_t(block_type_t::if_block); }
 
-block_t block_t::event_block(event_t evt) {
+block_t event_block(rust::Box<Event> evt) {
     block_t b{block_type_t::event};
-    b.event.reset(new event_t(std::move(evt)));
+    b.event = std::make_shared<rust::Box<Event>>(std::move(evt));
     return b;
 }
 
diff --git a/src/parser.h b/src/parser.h
index 0fbbd1a3d..66ede3464 100644
--- a/src/parser.h
+++ b/src/parser.h
@@ -25,7 +25,7 @@
 #include "util.h"
 #include "wait_handle.h"
 
-struct event_t;
+struct Event;
 class io_chain_t;
 class autoclose_fd_t;
 
@@ -64,11 +64,10 @@ enum class loop_status_t {
 
 /// block_t represents a block of commands.
 class block_t {
-   private:
+   public:
     /// Construct from a block type.
     explicit block_t(block_type_t t);
 
-   public:
     // If this is a function block, the function name. Otherwise empty.
     wcstring function_name{};
 
@@ -82,7 +81,7 @@ class block_t {
     filename_ref_t src_filename{};
 
     // If this is an event block, the event. Otherwise ignored.
-    std::shared_ptr<event_t> event;
+    std::shared_ptr<rust::Box<Event>> event;
 
     // If this is a source block, the source'd file, interned.
     // Otherwise nothing.
@@ -113,7 +112,6 @@ class block_t {
 
     /// Entry points for creating blocks.
     static block_t if_block();
-    static block_t event_block(event_t evt);
     static block_t function_block(wcstring name, wcstring_list_t args, bool shadows);
     static block_t source_block(filename_ref_t src);
     static block_t for_block();
@@ -127,6 +125,8 @@ class block_t {
     const event_blockage_list_t *ffi_event_blocks() const { return &event_blocks; }
 };
 
+block_t event_block(rust::Box<Event> evt);
+
 struct profile_item_t {
     using microseconds_t = long long;
 
@@ -491,9 +491,9 @@ class parser_t : public std::enable_shared_from_this<parser_t> {
 
     /// autocxx junk.
     RustFFIJobList ffi_jobs() const;
-    event_blockage_list_t &ffi_global_event_blocks() { return global_event_blocks; }
+    event_blockage_list_t *ffi_global_event_blocks() { return &global_event_blocks; }
     job_t *ffi_job_get_from_pid(int pid) const;
-    library_data_pod *ffi_libdata_pod() const;
+    library_data_pod *ffi_libdata_pod();
 
     ~parser_t();
 };
diff --git a/src/proc.cpp b/src/proc.cpp
index 53e4c8024..0911cfe1a 100644
--- a/src/proc.cpp
+++ b/src/proc.cpp
@@ -238,7 +238,7 @@ static void handle_child_status(const shared_ptr<job_t> &job, process_t *proc,
             if (is_interactive_session()) {
                 // Mark the job group as cancelled.
                 job->group->cancel_with_signal(sig);
-            } else if (!event_is_signal_observed(sig)) {
+            } else if (!is_signal_observed(sig)) {
                 // Deliver the SIGINT or SIGQUIT signal to ourself since we're not interactive.
                 struct sigaction act;
                 sigemptyset(&act.sa_mask);
@@ -463,33 +463,34 @@ static void process_mark_finished_children(parser_t &parser, bool block_ok) {
 }
 
 /// Generate process_exit events for any completed processes in \p j.
-static void generate_process_exit_events(const job_ref_t &j, std::vector<event_t> *out_evts) {
+static void generate_process_exit_events(const job_ref_t &j,
+                                         std::vector<rust::Box<Event>> *out_evts) {
     // Historically we have avoided generating events for foreground jobs from event handlers, as an
     // event handler may itself produce a new event.
     if (!j->from_event_handler() || !j->is_foreground()) {
         for (const auto &p : j->processes) {
             if (p->pid > 0 && p->completed && !p->posted_proc_exit) {
                 p->posted_proc_exit = true;
-                out_evts->push_back(event_t::process_exit(p->pid, p->status.status_value()));
+                out_evts->push_back(new_event_process_exit(p->pid, p->status.status_value()));
             }
         }
     }
 }
 
 /// Given a job that has completed, generate job_exit and caller_exit events.
-static void generate_job_exit_events(const job_ref_t &j, std::vector<event_t> *out_evts) {
+static void generate_job_exit_events(const job_ref_t &j, std::vector<rust::Box<Event>> *out_evts) {
     // Generate proc and job exit events, except for foreground jobs originating in event handlers.
     if (!j->from_event_handler() || !j->is_foreground()) {
         // job_exit events.
         if (j->posts_job_exit_events()) {
             auto last_pid = j->get_last_pid();
             if (last_pid.has_value()) {
-                out_evts->push_back(event_t::job_exit(*last_pid, j->internal_job_id));
+                out_evts->push_back(new_event_job_exit(*last_pid, j->internal_job_id));
             }
         }
     }
     // Generate caller_exit events.
-    out_evts->push_back(event_t::caller_exit(j->internal_job_id, j->job_id()));
+    out_evts->push_back(new_event_caller_exit(j->internal_job_id, j->job_id()));
 }
 
 /// \return whether to emit a fish_job_summary call for a process.
@@ -536,9 +537,8 @@ bool job_or_proc_wants_summary(const shared_ptr<job_t> &j) {
 
 /// Invoke the fish_job_summary function by executing the given command.
 static void call_job_summary(parser_t &parser, const wcstring &cmd) {
-    event_t event(event_type_t::generic);
-    event.desc.str_param1 = L"fish_job_summary";
-    block_t *b = parser.push_block(block_t::event_block(event));
+    auto event = new_event_generic(L"fish_job_summary");
+    block_t *b = parser.push_block(event_block(std::move(event)));
     auto saved_status = parser.get_last_statuses();
     parser.eval(cmd, io_chain_t());
     parser.set_last_statuses(saved_status);
@@ -667,7 +667,7 @@ static bool process_clean_after_marking(parser_t &parser, bool allow_interactive
 
     // Accumulate exit events into a new list, which we fire after the list manipulation is
     // complete.
-    std::vector<event_t> exit_events;
+    std::vector<rust::Box<Event>> exit_events;
 
     // Defer processing under-construction jobs or jobs that want a message when we are not
     // interactive.
@@ -719,7 +719,7 @@ static bool process_clean_after_marking(parser_t &parser, bool allow_interactive
 
     // Post pending exit events.
     for (const auto &evt : exit_events) {
-        event_fire(parser, evt);
+        event_fire(parser, *evt);
     }
 
     if (printed) {
diff --git a/src/signals.cpp b/src/signals.cpp
index 5b91d5e8a..acf4fd4e1 100644
--- a/src/signals.cpp
+++ b/src/signals.cpp
@@ -224,9 +224,9 @@ static void fish_signal_handler(int sig, siginfo_t *info, void *context) {
     }
 
     // Check if fish script cares about this.
-    const bool observed = event_is_signal_observed(sig);
+    const bool observed = is_signal_observed(sig);
     if (observed) {
-        event_enqueue_signal(sig);
+        enqueue_signal(sig);
     }
 
     // Do some signal-specific stuff.
diff --git a/src/termsize.cpp b/src/termsize.cpp
index 8bdcd0d2a..75ba9685b 100644
--- a/src/termsize.cpp
+++ b/src/termsize.cpp
@@ -50,6 +50,10 @@ termsize_container_t &termsize_container_t::shared() {
     return *res;
 }
 
+termsize_t termsize_container_t::ffi_updating(parser_t &parser) {
+    return shared().updating(parser);
+}
+
 termsize_t termsize_container_t::data_t::current() const {
     // This encapsulates our ordering logic. If we have a termsize from a tty, use it; otherwise use
     // what we have seen from the environment.
```

Reply:
Thanks, I've worked this in, but `event_block` is still broken and I have no idea how to get cxx and autocxx to cooperate there.